### DISCUSSION TITLE
CADViewer in CSharpCAD of possible interest to JSCAD.

### DISCUSSION THREAD

**Discussion by briansturgill** [jscad/OpenJSCAD.org#1125] [Show and tell]
Basically, it is a pyvista-based viewer. It has an HTTP-POST protocol that allow you to send a mesh for viewing.
It works like this:

The first time a CSharpCAD/PyCSCAD program sends a mesh to view, it first sends a "clear" that clears out previous models.
For each call to "view" or "save", a mesh is sent.
View is designed to be easy to insert... for example

function....
...
return geom3

Can have view inserted like this:
return view(geom3)

The point is you get a series of meshes sent from your running program that you can browse and manipulate.
It makes debugging much easier than any CAD environment I've seen.
(Geom2's are first extruded very thinly, and sent as a 3D mesh.)

It would be easy to do this in JSCAD as it's just an http post... you can see the protocol here:
https://github.com/briansturgill/CSharpCAD/blob/main/CSharpCAD/src/misc/View.cs

You can find CADViewer here:
https://github.com/briansturgill/CSharpCAD/tree/main/CADViewer

I consider it a bit unpolished, but it is great as a proof of concept and I'm already finding it very handy for mesh debugging.

**Comment by briansturgill** [comment]
Oh, if you just want to play around with it, the Python API is sufficiently ready...
See: https://github.com/briansturgill/CSharpCAD/tree/main/Python
Don't forget to install dotnet first: https://docs.microsoft.com/en-us/dotnet/core/install/
But it will keep you from having to deal with C# if your not already up to speed on it.

**Reply by z3dev** [reply]
@briansturgill this doesn't seem like a discussion, but rather information about your project. I'm going to drop this discussion if there's nothing more. (FYI, I don't know of another way to archive / complete a discussion, which seems kind of lame.)

**Reply by hrgdavor** [reply]
@z3dev I like @briansturgill insights about his project. I do not mid this info being here. I have changed category to show and tell.

**Reply by briansturgill** [reply]
@z3dev Please reread my post. "It would be easy to do this in JSCAD as it's just an http post... you can see the protocol here:
https://github.com/briansturgill/CSharpCAD/blob/main/CSharpCAD/src/misc/View.cs"
CADViewer was created because I needed a viewer for CSharpCAD... but I left it general. It can easily work with JSCAD.
Just take a look at the View.cs link above -- the whole thing is just 144 lines.
I purposely designed the protocol to be friendly to JSCAD.
It's intended for people that use their CAD tool from the command line (usually so they can use a decent editor) but of course want to see their output (hopefully easily and quickly.

When I was using OpenSCAD, it's lack of being able to see more than one object at a time (all objects ended up in a common blob).
Likewise it had no good way of seeing individual pieces simultaneously. At that time it also had only one way to save objects in multiple files and that essentially meant rebuilding the whole set of models for each object you wanted to save.
Thus, I have a View and a Save (save also does a view so I can check it out without bringing up a slicer).

Anyway, this EASILY works with JSCAD. The current viewer is basically just something I threw together as a proof of concept.
If you want to do this in JSCAD, I'd be delighted to cooperate. If the protocol doesn't suit you, I'm willing to make changes.

I was making an offer of working with you, not showing off my project. My project is not yet ready for showing off.
@hrgdavor Thanks for pointing out show and tell, when I do get ready to show off my project, I'll post there.

Anyway, it is liberating to have the ability to take arbitrary 3D-manipulatable views when developing models.
That is what CADViewer is all about.

If you don't make the js view, I probably will as it would be nice when I'm trying things in JSCAD.
No time at the moment... busy working on 3D booleans.

**Reply by hrgdavor** [reply]
@briansturgill js in browser can not listen to HTTP, but websocket could be used. By adding a listen button, js could poll a websocket port constantly, and the commandline can then start websocket server on that port when neeeded and push things.

for this purpose a commandline tool could be made that pushes input to jscad. That way the command line users need not know how to use websockets, but can just `$ command_produce_sth_for_jscad.sh > jscadws -p 23453 `

... that said it may be an overkill ... I have an idea that works right now

if you have a commandline that produces sth for jscad  `$ command_produce_sth_for_jscad.sh > test.js `

then you open `https://openjscad.xyz/` and turn on `Auto reload` and drag and drop the `test.js`

each time you want to update the output you can run  `$ command_produce_sth_for_jscad.sh > test.js `
and jscad will notice changed file and reload it.

**Reply by briansturgill** [reply]
@hrgdavor I'm not sure I am following. CADViewer has an internal (localhost only) web browser. To view something, one simple posts to that server, which can be done from browser js as well as node js. I designed that way intentionally -- I mean what's the point of JSCAD if it doesn't run in the browser?

**Reply by hrgdavor** [reply]
@briansturgill I am saying that HTTP POST is not necessary for jscad, it can watch for file changes. To view something, one simple changes the watched file :)

To me personally JSCAD is not only about the browser. The point of JSCAD for me is that I prefer JavaScript over all other lanaguages for UI and in this instance for code-cad. The reason I prefer JavaScript is that I like coding in it, and can run JS apps in many environments, not just browser, and of course browser based UI on many platforms (desktops, mobile, ...).

**Reply by briansturgill** [reply]
> @briansturgill I am saying that HTTP POST is not necessary for jscad, it can watch for file changes. To view something, one simple changes the watched file :)
...

Okay, but you often want to see more than one view. For example, let's say I making a "project box" for a raspberry pi project.
I will have a box and a lid. Both pieces will have complicated holes and posts added to hold parts in place.
It is vitally important to be able to compare the lid and box, yet still independently rotate them for checking.
CADViewer makes this easy.

Another example... most components that are 3D printed have sub components. With "view" one can see each part at each step.
(One can switch between them with arrow keys.) This makes model design much, much easier.
CADViewer has bounding rulers for measurement, ability to toggle wireframe and points views and it works with the ease of console.log.

**Reply by hrgdavor** [reply]
@briansturgill those are all nice features. It would be nice to see them is some docs with screenshots. I am way behind on things I want to do in JSCAD and it will be some time before I can take time to run your viewer and consider integrating it (or copy ideas if better suited).

One very major thing is the new jscad prototype, because to consider any new features I need a JSCAD UI that I am comfortable changing and maintaining, and also any such new version of jscad needs to be something that also @z3dev would be comfortable working with.

the list of things I bookmarked and would like to explore in the scope of JSCAD is so long I will likely not explore it all before my retirement ....

![image](https://user-images.githubusercontent.com/2480762/186505982-2c4b7f56-0928-4bac-96de-f8517fad21e9.png)
